### PR TITLE
Pass Credentials through GrpcClient

### LINKED ISSUE #31: Take credential creation out of GrpcClient
Per Dan's [comment](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/pull/26#discussion_r281736123) on #26.

Credential creation can cause a confounding condition (yes, I am just here for the alliteration. I'll show myself out) because of IOException thrown when searching for credentials on file system. A provider will give us flexibility to swap out credential strategies, and would be a good place to corral the checked exception.

### PR DESCRIPTION
**Pull request by dzou**
Refactor GrpcClient to accept credentials through the constructor.

Fixes #31.

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 1]
```diff
@@ -28,22 +32,37 @@
...
    * @param instanceName instance to connect to
    * @param databaseName database to connect to.
+   * @param credentialsLocation GCP credentials to authenticate service calls with.
    */
   private SpannerConnectionConfiguration(
-      String projectId, String instanceName, String databaseName) {
+      String projectId,
+      String instanceName,
+      String databaseName,
+      @Nullable String credentialsLocation) {
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: If it's optional, either use a setter or overload the constructor.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java [hunk 2]
```diff
@@ -28,22 +32,37 @@
...
     Assert.requireNonNull(databaseName, "databaseName must not be null");
 
     this.fullyQualifiedDbName = String.format(
         FULLY_QUALIFIED_DB_NAME_PATTERN, projectId, instanceName, databaseName);
+    this.credentialsLocation = credentialsLocation;
+  }
+
+  GoogleCredentials createCredentials() throws IOException {
+    if (credentialsLocation != null) {
+      return GoogleCredentials.fromStream(new FileInputStream(credentialsLocation));
```
Comment (meltsufin) [inline src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: In Spring Cloud GCP, we also support providing the `encodedKey` directly, which is useful when one is storing the credentials somewhere other than a file.
In any case, why not just allow the user to pass in `GoogleCredentials` object, which they can build however they like.

Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: I agree. It's best for the driver to not try to infer state from the environment. We can always create a starter/autoconfiguration in `spring-cloud-gcp` to support common usages for the driver.

**Comment by elefeint** [comment]
I agree with injecting credentials instead of figuring them out in the driver.
Since `ConnectionFactoryOptions` can be of any type, the end user can directly pass credentials when creating a connection factory.

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 1 above)
Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: this is a private constructor, though. The actual work is done in the builder.
I suppose we can have a private setter, but it makes the object not completely immutable.

Comment (meltsufin) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: fair point!

File: src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java (hunk 2 above)
Comment (meltsufin) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: I think using the application default credentials when the user doesn't specify the credentials object is OK. It's a really nice feature that's available across the board on GCP.

Comment (elefeint) [reply src/main/java/com/google/cloud/spanner/r2dbc/SpannerConnectionConfiguration.java]: Hmm, that's true, and that only requires a dependency on google-auth (although the dependency name `google-auth-library-oauth2-http` might be a tad confusing in the context of non-http application. I wonder if there is a leaner dependency we could use.

**Pull request merged**